apps/emqx_durable_storage/src/emqx_ds_storage_skipstream_lts.erl

```diff
@@ -0,0 +1,739 @@
+    ok = rocksdb:drop_column_family(DBHandle, DataCF),
+    ok = rocksdb:drop_column_family(DBHandle, TrieCF),
+    ok.
+
+prepare_batch(
+    _ShardId,
+    S = #s{trie = Trie, hash_bytes = HashBytes},
+    Messages,
+    _Options
+) ->
+    _ = erase(?lts_persist_ops),
+    Payloads =
+        lists:flatmap(
+            fun({Timestamp, Msg = #message{topic = Topic}}) ->
+                Tokens = words(Topic),
+                {Static, Varying} = emqx_ds_lts:topic_key(Trie, fun threshold_fun/1, Tokens),
+                %% TODO: is it possible to create index during the
+                %% commit phase to avoid transferring indexes through
+                %% the translog?
+                [
+                    {mk_key(Static, 0, <<>>, Timestamp), serialize(S, Varying, Msg)}
+                    | mk_index(HashBytes, Static, Timestamp, Varying)
+                ]
+            end,
+            Messages
+        ),
+    {ok, #{
+        ?cooked_payloads => Payloads,
+        ?cooked_lts_ops => pop_lts_persist_ops()
+    }}.
+
+commit_batch(
+    _ShardId,
+    #s{db = DB, trie_cf = TrieCF, data_cf = DataCF, trie = Trie},
+    #{?cooked_lts_ops := LtsOps, ?cooked_payloads := Payloads},
+    Options
+) ->
+    {ok, Batch} = rocksdb:batch(),
+    try
+        %% Commit LTS trie to the storage:
+        lists:foreach(
+            fun({Key, Val}) ->
+                ok = rocksdb:batch_put(Batch, TrieCF, term_to_binary(Key), term_to_binary(Val))
+            end,
+            LtsOps
+        ),
+        %% Apply LTS ops to the memory cache:
+        _ = emqx_ds_lts:trie_update(Trie, LtsOps),
+        %% Commit payloads:
+        lists:foreach(
+            fun({Key, Val}) ->
+                ok = rocksdb:batch_put(Batch, DataCF, Key, Val)
+            end,
+            Payloads
+        ),
+        Result = rocksdb:write_batch(DB, Batch, [
+            {disable_wal, not maps:get(durable, Options, true)}
+        ]),
+        %% NOTE
+        %% Strictly speaking, `{error, incomplete}` is a valid result but should be impossible to
+        %% observe until there's `{no_slowdown, true}` in write options.
+        case Result of
+            ok ->
+                ok;
+            {error, {error, Reason}} ->
+                {error, unrecoverable, {rocksdb, Reason}}
+        end
+    after
+        rocksdb:release_batch(Batch)
+    end.
+
+get_streams(_Shard, #s{trie = Trie}, TopicFilter, _StartTime) ->
+    get_streams(Trie, TopicFilter).
+
+get_delete_streams(_Shard, #s{trie = Trie}, TopicFilter, _StartTime) ->
+    get_streams(Trie, TopicFilter).
+
+make_iterator(_Shard, #s{trie = Trie}, #stream{static_index = StaticIdx}, TopicFilter, StartTime) ->
+    {ok, TopicStructure} = emqx_ds_lts:reverse_lookup(Trie, StaticIdx),
+    CompressedTF = emqx_ds_lts:compress_topic(StaticIdx, TopicStructure, TopicFilter),
+    {ok, #it{
+        static_index = StaticIdx,
+        ts = StartTime,
+        compressed_tf = emqx_topic:join(CompressedTF)
```

Comment:
Nit. It kind of seems like `compressed_tf` is used in "tokenized" form much more often in this module.

Reply:
Erlang will employ TLV encoding to encode list of binaries, which has overhead. For example:
  
```
5> size(term_to_binary([<<"a">>, <<"b">>])).
19
6> size(term_to_binary(<<"a/b">>)).
9
```

Given that iterator objects are stored and transfered over network, it's _a_ factor. There's no way to predict what cost will be larger:

- of parsing the topic
- of reading and writing larger iterator objects from disk and sending them over network.

So I chose to join them for easier debugging, I guess.